File: src/value.zig
const std = @import("std");
const memory = @import("memory.zig");
const Allocator = std.mem.Allocator;

pub const Value = f64;

pub const ValueArray = struct {
    const Self = @This();

    capacity: usize,
    count: usize,
    data: ?[]Value,
    allocator: Allocator,

    pub fn init(allocator: Allocator) Self {
        return Self{
            .allocator = allocator,
            .capacity = 0,
            .count = 0,
            .data = null,
        };
    }
    pub fn free(self: *Self) void {
        _ = try memory.freeArray(Value, self.allocator, self.data, self.capacity);
        self.capacity = 0;
        self.count = 0;
        self.data = null;
    }

    pub fn write(self: *Self, item: Value) !void {
        if (self.count >= self.capacity) {
            const old_capacity = self.capacity;
            self.capacity = memory.growCapacity(old_capacity);
            self.data = try memory.growArray(Value, self.allocator, self.data, old_capacity, self.capacity);
        }
        self.data.?[self.count] = item;
        self.count += 1;
    }
};

pub fn printValue(value: Value) void {
    std.debug.print("{d}", .{value});
}

File: c/src/value.c
#include <stdio.h>

#include "memory.h"
#include "value.h"

void ValueArray_init(ValueArray* array) {
    array->values = NULL;
    array->capacity = 0;
    array->count = 0;
}

void ValueArray_write(ValueArray* array, Value value) {
    if (array->capacity < array->count + 1) {
        int old_capacity = array->capacity;
        array->capacity = GROW_CAPACITY(old_capacity);
        array->values = GROW_ARRAY(Value, array->values, old_capacity, array->capacity);
    }
    array->values[array->count] = value;
    array->count++;
}

void ValueArray_free(ValueArray* array) {
    FREE_ARRAY(Value, array->values, array->capacity);
    ValueArray_init(array);
}

void printValue(Value value) {
    printf("%g", value);
}
